File: code/zato-web-admin/src/zato/admin/web/views/outgoing/wsx.py
# -*- coding: utf-8 -*-

"""
Copyright (C) 2023, Zato Source s.r.o. https://zato.io

Licensed under AGPLv3, see LICENSE.txt for terms and conditions.
"""

# stdlib
import logging

# Zato
from zato.common.api import GENERIC
from zato.admin.web.forms import ChangePasswordForm
from zato.admin.web.forms.outgoing.wsx import CreateForm, EditForm
from zato.admin.web.views import CreateEdit, Delete as _Delete, Index as _Index
from zato.common.odb.model import GenericConn

# ################################################################################################################################

logger = logging.getLogger(__name__)

# ################################################################################################################################

class Index(_Index):
    method_allowed = 'GET'
    url_name = 'out-wsx'
    template = 'zato/outgoing/wsx.html'
    service_name = 'zato.generic.connection.get-list'
    output_class = GenericConn
    paginate = True

    class SimpleIO(_Index.SimpleIO):
        input_required = ('cluster_id', 'type_')
        output_required = ('id', 'name', 'address', 'address_masked')
        output_optional = ('is_active', 'is_zato',
            'on_connect_service_id', 'on_connect_service_name', 'on_message_service_id', 'on_message_service_name',
            'on_close_service_id', 'on_close_service_name', 'subscription_list', 'security_def', 'has_auto_reconnect',
            'data_format', 'ping_interval', 'pings_missed_threshold', 'socket_read_timeout', 'socket_write_timeout')
        output_repeated = True

    def handle(self):
        if self.req.zato.get('client'):
            security_list = self.get_sec_def_list('basic_auth').def_items
        else:
            security_list = []

        return {
            'create_form': CreateForm(security_list, req=self.req),
            'edit_form': EditForm(security_list, prefix='edit', req=self.req),
            'change_password_form': ChangePasswordForm()
        }

# ################################################################################################################################

class _CreateEdit(CreateEdit):
    method_allowed = 'POST'

    class SimpleIO(CreateEdit.SimpleIO):
        input_required = ('name', 'is_active', 'is_zato', 'address', 'on_connect_service_name', 'on_message_service_name',
            'on_close_service_name', 'subscription_list', 'security_def', 'has_auto_reconnect', 'data_format',
            'ping_interval', 'pings_missed_threshold', 'socket_read_timeout', 'socket_write_timeout')
        output_required = ('id', 'name')

    def populate_initial_input_dict(self, initial_input_dict):
        initial_input_dict['type_'] = GENERIC.CONNECTION.TYPE.OUTCONN_WSX
        initial_input_dict['is_internal'] = False
        initial_input_dict['is_channel'] = False
        initial_input_dict['is_outconn'] = True
        initial_input_dict['pool_size'] = 1
        initial_input_dict['sec_use_rbac'] = False

    def post_process_return_data(self, return_data):

        for name in ('connect', 'message', 'close'):
            field_name = 'on_{}_service_name'.format(name)
            return_data[field_name] = self.input_dict.get(field_name)

        return return_data

    def success_message(self, item):
        return 'Successfully {} outgoing WebSocket connection `{}`'.format(self.verb, item.name)

# ################################################################################################################################

class Create(_CreateEdit):
    url_name = 'out-wsx-create'
    service_name = 'zato.generic.connection.create'

# ################################################################################################################################

class Edit(_CreateEdit):
    url_name = 'out-wsx-edit'
    form_prefix = 'edit-'
    service_name = 'zato.generic.connection.edit'

# ################################################################################################################################

class Delete(_Delete):
    url_name = 'out-wsx-delete'
    error_message = 'Could not delete outgoing WebSocket connection'
    service_name = 'zato.generic.connection.delete'

# ################################################################################################################################

File: code/zato-web-admin/src/zato/admin/static/js/outgoing/wsx.js

// /////////////////////////////////////////////////////////////////////////////

$.fn.zato.data_table.WSX = new Class({
    toString: function() {
        var s = '<WSX id:{0} name:{1} is_active:{2}';
        return String.format(s, this.id ? this.id : '(none)',
                                this.name ? this.name : '(none)',
                                this.is_active ? this.is_active : '(none)');
    }
});

// /////////////////////////////////////////////////////////////////////////////

$(document).ready(function() {
    $('#data-table').tablesorter();
    $.fn.zato.data_table.password_required = true;
    $.fn.zato.data_table.class_ = $.fn.zato.data_table.WSX;
    $.fn.zato.data_table.new_row_func = $.fn.zato.outgoing.wsx.data_table.new_row;
    $.fn.zato.data_table.parse();
    $.fn.zato.data_table.setup_forms(['name', 'address', 'security_def']);
})

$.fn.zato.outgoing.wsx.create = function() {
    $.fn.zato.data_table._create_edit('create', 'Create a new outgoing WebSocket connection', null);
}

$.fn.zato.outgoing.wsx.edit = function(id) {
    $.fn.zato.data_table._create_edit('edit', 'Update the outgoing WebSocket connection', id);
}

$.fn.zato.outgoing.wsx.data_table.new_row = function(item, data, include_tr) {
    var row = '';

    if(include_tr) {
        row += String.format("<tr id='tr_{0}' class='updated'>", item.id);
    }

    var is_active = item.is_active == true;
    var is_zato = item.is_zato == true;
    var has_auto_reconnect = item.has_auto_reconnect == true;
    var data_format = item.data_format || '';

    /* 1 */
    row += "<td class='numbering'>&nbsp;</td>";
    row += "<td class='impexp'><input type='checkbox' /></td>";
    row += String.format('<td>{0}</td>', item.name);

    /* 2 */
    row += String.format('<td>{0}</td>', is_active ? 'Yes' : 'No');
    row += String.format('<td>{0}</td>', item.address_masked);
    row += String.format('<td>{0}</td>', is_zato ? 'Yes' : 'No');

    /* 3 a */
    if(data.on_connect_service_name) {
        row += String.format('<td>{0}</td>',
            String.format("<a href='/zato/service/overview/{0}/?cluster={1}'>{0}</a>",
            data.on_connect_service_name, item.cluster_id));
    }
    else {
        row += $.fn.zato.empty_table_cell;
    }

    /* 3 b */
    if(data.on_message_service_name) {
        row += String.format('<td>{0}</td>',
            String.format("<a href='/zato/service/overview/{0}/?cluster={1}'>{0}</a>",
            data.on_message_service_name, item.cluster_id));
    }
    else {
        row += $.fn.zato.empty_table_cell;
    }

    /* 3 c */
    if(data.on_close_service_name) {
        row += String.format('<td>{0}</td>',
            String.format("<a href='/zato/service/overview/{0}/?cluster={1}'>{0}</a>",
            data.on_close_service_name, item.cluster_id));
    }
    else {
        row += $.fn.zato.empty_table_cell;
    }

    /* 4 */
    row += String.format('<td>{0}</td>', String.format("<a href=\"javascript:$.fn.zato.outgoing.wsx.edit('{0}')\">Edit</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.outgoing.wsx.delete_({0});'>Delete</a>", item.id));
    row += String.format("<td class='ignore item_id_{0}'>{0}</td>", item.id);

    /* 5 */
    row += String.format("<td class='ignore'>{0}</td>", is_active);
    row += String.format("<td class='ignore'>{0}</td>", is_zato);
    row += String.format("<td class='ignore'>{0}</td>", item.on_connect_service_name);

    /* 6 */
    row += String.format("<td class='ignore'>{0}</td>", item.on_message_service_name)
    row += String.format("<td class='ignore'>{0}</td>", item.on_close_service_name);
    row += String.format("<td class='ignore'>{0}</td>", item.security_def);

    /* 7 */
    row += String.format("<td class='ignore'>{0}</td>", item.subscription_list);
    row += String.format("<td class='ignore'>{0}</td>", has_auto_reconnect);
    row += String.format("<td class='ignore'>{0}</td>", data_format);

    /* 8 */
    row += String.format("<td class='ignore'>{0}</td>", item.ping_interval);
    row += String.format("<td class='ignore'>{0}</td>", item.pings_missed_threshold);
    row += String.format("<td class='ignore'>{0}</td>", item.socket_read_timeout);

    /* 9 */
    row += String.format("<td class='ignore'>{0}</td>", item.socket_write_timeout);
    row += String.format("<td class='ignore'>{0}</td>", item.address);

    if(include_tr) {
        row += '</tr>';
    }

    return row;
}

$.fn.zato.outgoing.wsx.delete_ = function(id) {
    $.fn.zato.data_table.delete_(id, 'td.item_id_',
        'Outgoing WebSocket connection `{0}` deleted',
        'Are you sure you want to delete outgoing WebSocket connection `{0}`?',
        true);
}
